arraycontext/parameter_study/transform.py

```diff
@@ -0,0 +1,602 @@
+
+        for ind, array in enumerate(new_arrays):
+            for _, axis in enumerate(array.axes):
+                axis_tags = list(axis.tags_of_type(ParameterStudyAxisTag))
+                if axis_tags:
+                    assert len(axis_tags) == 1
+                    new_access_descriptors[ind] = new_access_descriptors[ind] + \
+                                                (EinsumElementwiseAxis(dim=study_to_axis_number[axis_tags[0]]),)
+
+        return Einsum(tuple(new_access_descriptors), new_arrays,
+                     axes=expr.axes + new_axes_for_end,
+                     redn_axis_to_redn_descr=expr.redn_axis_to_redn_descr,
+                     index_to_access_descr=expr.index_to_access_descr,
+                     tags=expr.tags,
+                     non_equality_tags=expr.non_equality_tags)
+
+    # }}} Operations with multiple predecessors.
+
+
+class ParamAxisExpander(IdentityMapper):
+
+    def map_subscript(self, expr: prim.Subscript,
+                      varname_to_studies: Mapping[str,
+                                                  Mapping[ParameterStudyAxisTag, bool]],
+                      study_to_axis_number: Mapping[ParameterStudyAxisTag, int]):
+        # We know that we are not changing the variable that we are indexing into.
+        # This is stored in the aggregate member of the class Subscript.
+
+        # We only need to modify the indexing which is stored in the index member.
+        name = expr.aggregate.name
+        if name in varname_to_studies.keys():
+            #  These are the single instance information.
+            index = self.rec(expr.index, varname_to_studies,
+                             study_to_axis_number)
+
+            new_vars: Tuple[prim.Variable, ...] = ()
+
+            for key, num in sorted(study_to_axis_number.items(),
+                                   key=lambda item: item[1]):
+                if key in varname_to_studies[name]:
+                    new_vars = *new_vars, prim.Variable(f"_{num}"),
+
+            if isinstance(index, tuple):
+                index = index + new_vars
+            else:
+                index = tuple(index) + new_vars
+            return type(expr)(aggregate=expr.aggregate, index=index)
+        return expr
+
+
+# {{{ ParamStudyPytatoPyOpenCLArrayContext
+
+
+class ParamStudyPytatoPyOpenCLArrayContext(PytatoPyOpenCLArrayContext):
+    """
+    A derived class for PytatoPyOpenCLArrayContext updated for the
+    purpose of enabling parameter studies and uncertainty quantification.
+
+    .. automethod:: __init__
+
+    .. automethod:: transform_dag
+
+    .. automethod:: compile
+    """
+
+    def compile(self, f: Callable[..., Any]) -> Callable[..., Any]:
+        return ParamStudyLazyPyOpenCLFunctionCaller(self, f)
+
+    def transform_loopy_program(self,
+                                t_unit: lp.TranslationUnit) -> lp.TranslationUnit:
+        # Update in a subclass if you want.
+        return t_unit
```

Comment:
This is here so that I can actually execute the code made in the examples folder. Perhaps it would be better to just make a subclass in the examples code and then use that instead.